README.md
# P2N-v3 - docker version
     _____      _             _     ___    _   _      _           _____   ___    _   _ 
    |  __ \    | |           | |   |__ \  | \ | |    | |      /  |  __ \ |__ \  | \ | | \
    | |__) |_ _| |_ ___ _ __ | |_     ) | |  \| | ___| |_    /   | |__) |   ) | |  \| |  \
    |  ___/ _` | __/ _ \ '_ \| __|   / /  | . ` |/ _ \ __|   |   |  ___/   / /  | . ` |   |
    | |  | (_| | ||  __/ | | | |_   / /_  | |\  |  __/ |_     \  | |      / /_  | |\  |  /
    |_|   \__,_|\__\___|_| |_|\__| |____| |_| \_|\___|\__|     \ |_|     |____| |_| \_| /       

What's new ?
-----
* Patent2Net now works in Docker. To install it simply download "Docker_P2N.zip" and follow the "Readme.txt"
* DO NOT INSTALL AS IT on the cloud: care to unactivate FTP feature.
* Patent2Net now integrate an user interface to enter request instead of creating cql file.
* When executing setup.py, P2N will install Flask by pip command. (needed for the user interface)
* To access it you need to launch app.py and go to 127.0.0.1:5000 with your web browser
* DATA stuff can be acceded by FTP (Filezila on localhost for instance).
* Update is operated using the "UPDATE" button. Silly, ins't it?
 

About
-----
[Patent2Net] (http://patent2netv2.vlab4u.info/dokuwiki/doku.php?id=page) is :
* elaborated and maintained (on a free base) by a [small international team] (http://patent2netv2.vlab4u.info/dokuwiki/doku.php?id=about_p2n:community;) of university professors and researchers.  
* an "open source" package and contributions are welcome
* available "[as it is] (http://patent2netv2.vlab4u.info/dokuwiki/doku.php?id=about_p2n:disclaimer)".

Patent2Net is a free package, dedicated to:
* permit the use of patent information in academic, nano and small firms, developing countries;
* learn, study and practice how to collect, treat and communicate "textual bibliographic information", and automation process;
* provide statistical analysis and representations of a set of patents;
* create research corpora (of patents)

The results of statistical patents indicators and text processing stuff can be explored with the firefox browser on your localhost or exported to your own website like here: [http://patent2netv2.vlab4u.info/](http://patent2netv2.vlab4u.info/)



CHANGES.rst
.. _Changes:

*****************************
Patent2Net CHANGES (in short)
*****************************

##################
Patent2Net CHANGES
##################

2021-01-01 4.0.0-dev1
=====================
.. note::
 - Major improvements in the global architecture
 - P2N in now built on centOS v8
 - Added a docker image for Elasticsearch: 7.9.2 with carrot2 plugin (see https://github.com/carrot2/elasticsearch-carrot2). localhost:9200 and 9300 to access the Elasticsearch server
 - Added a docker image for Kibana: 7.9.2 (see https://www.elastic.co/fr/kibana). Also in the standard port 5601.
 - First version of P2N indexer that feed the elastic search indexer
.. warning:: 
 - ElasticSearch and Kibana are just here to start the new ways for P2N. At this time no features are provided at all. Specialists are welcome
 - Secure version of VSFTPD, fixed several security bugs (port 20-21). 

.. warning:: 
 User p2n with no passwd is set in Dockerfile. Mind in changing this if installing P2N on the cloud!


2020-11-30 3.1.0-dev5
=====================
- New processing feature in beta version (thanks Nezha): Trizifyer
- Major improvement in time processing in network production (using new libraries)
- but dynamic network aren't functional yet. Imrpove this

2020-07-20 3.1.0-dev4
=====================
- New features of "progess bar"
.. note::
 Help to fix the ugly page. UX and designer skills welcome!
- SSE system for all data processing scripts

2020-07-20 3.1.0-dev3
=====================
- Introduction of new libraries in various processing steps: pandas, mpld3, scipy
- New processing steps of data consolidation:
 - Applicants and Inventor Names normalisation (integration of a table from EPO with more than a million of entries)
 - Added more thant 2000 lines in previous file thank to URFIST project
 - New filtering process. Patents list collected from a request may contain (or not) equivalents patent. This tend to make biaises in several processing features of P2N. The new process extracts in the dataset the oldest patent in each subfamilies encountered in the set.
- several bug fixes
- added VSFTPD feature to communicate with the docker image

2020-06-20 3.1.0-dev2
=====================
- New branch: docker installation
- P2N works now in a standalone docker Ubuntu machine (Thanks Esteban).
 - Add Flask front end, localhost:5000 is the entry point.
 - Add update feature making p2n version allways uptodate. Static versions are useless from now.
 - Integration of several features (Pivot, Datatable)
 - Others exports (to IramuteQ or Carrot2) have to use the download data pages

2020-01-20 3.1.0-dev1
=====================
- firt publication of the automatic request splitter: a tool to skirt the 2000 limit of the EPO API
- various updates

2019 3.1.0-dev0
=====================

- various attemps in simplication of installation procedure
- many general bug fixes



development
===========
- Use ImageMagick for thumbnail generation in ``FusionImages.py``,
  see also https://github.com/Patent2net/P2N/issues/24
- Use "Pillow" for thumbnail generation, gracefully fall back to ImageMagick's "convert"

2018-03-20 3.0.0-dev6
=====================
- Make ``p2n adhoc dump`` output results either in "OpsExchangeDocument" or "Patent2NetBrevet" format
- Make ``p2n adhoc list`` output arbitrary fields of "OpsExchangeDocument" (projection)
- Improve documentation regarding CQL query language and updated command line parameters
- Introduce ad-hoc mode for generating JSON data suitable for
  feeding into PivotTable.js with Patent2NetBrevet data model
- Fix ``p2n adhoc dump`` with ``--with-register``
- Improve decoding raw JSON data into "OpsExchangeDocument" object instances
- Improve documentation
- Remove custom helper "attr_object_as_dict" in favor of vanilla "attr.asdict"
- Remove redundant "designated_states" field by using dotted name resolver
  for accessing "register.designated_states" when generating worldmap data
- Add support for PivotTable data generation in ad-hoc mode with "OPSExchangeDocument" data model
- Refactor maps.py and tables.py to "p2n.formatter" namespace
- Improve decoding of "OPSRegisterDocument"
- Improve error logging when decoding register information
- Decode historical data from register information
- Allow dotted attribute/field access for ``p2n adhoc list``
- Add advanced example about using ``jq`` to filter and reformat register information with ``p2n adhoc dump``
- Add convenience command ``p2n images`` for image data acquisition
- Properly display OPS error responses to the user, e.g. when
  submitting an invalid expression like ``--expression='foo=bar'``
- Add command ``p2n adhoc search`` for displaying search results only in raw OPS format
- Complete the data model for OPS Exchange Document
- Improve FusionImages.py
    - Add link to general template, fix and add context info for images
    - Image fusion
    - Improve layout/pagination for patent images gallery
    - Fix output thumbs
    - Add page size handler
- Improve gracefulness for Interface2.py and OPSGatherContentsV2-Images.py, see also #23.
- Fix ``p2n --version`` output, addressing #23. Thanks, @smorvan!

2017-12-01 3.0.0-dev5
=====================
- Attempt to add missing NameCountryMap.csv by providing MANIFEST.in file

2017-12-01 3.0.0-dev4
=====================
- Fix setup.py

2017-12-01 3.0.0-dev3
=====================
- Improve logging and error handling for register data acquisition

2017-12-01 3.0.0-dev2
=====================
- Add crawling behavior to new data acquisition subsystem
  to collect all results from OPS published data search
- Start project documentation based on Sphinx
- Improve documentation layout and move towards reStructuredText

2017-11-30 3.0.0-dev1
=====================
- Minor fixes re. argument processing
- Improve robustness re. case-sensitivity at map resource acquisition (countries.json)
- Add setup.py, convenience step runner and documentation

p2n/util.py
# -*- coding: utf-8 -*-
# (c) 2017-2018 The Patent2Net Developers
import os
import sys
import time
import types
import where
import logging
import operator
import functools
import itertools
import traceback
import subprocess
from six import StringIO, BytesIO
from json.encoder import JSONEncoder
from collections import OrderedDict

#########
# this is added because I don't know haow to import function from Patent2NNet Librairies (patent2Net\P2N_lib)
import requests

def AnnonceProgres(Appli, valActu, valMax):
    valActu = "%.2f" % valActu 
    pipo = requests.get('http://localhost:5000/announce?appli=%s&ValActu=%s&valMax=%s' %(Appli, valActu, valMax) )

# so I copy it here  FIX It please
#




logger = logging.getLogger(__name__)


def setup_logging(level=logging.INFO):
    log_format = '%(asctime)-15s [%(name)-25s] %(levelname)-7s: %(message)s'
    logging.basicConfig(
        format=log_format,
        stream=sys.stderr,
        level=level)


def boot_logging(options=None):
    log_level = logging.INFO
    if options and options.get('--debug'):
        log_level = logging.DEBUG
    setup_logging(level=log_level)


def normalize_docopt_options(options):
    normalized = {}
    for key, value in options.items():
        key = key.strip('--<>')
        normalized[key] = value
    return normalized


def run_script(script, configfile, directory='Patent2Net'):

    # Compute command
    command = 'python {script} {configfile}'.format(script=script, configfile=configfile)
    logger.info('Running command "{}"'.format(command))

    # Run process
    process = subprocess.Popen(command, shell=True, bufsize=1, cwd=directory)

    # Poll process for new output until finished
    while True:

        try:
            outcome = process.poll()

            if outcome is not None:
                returncode = process.returncode
                if returncode != 0:
                    logger.error(
                        'Command "{command}" failed with return code {returncode}'.format(**locals()))
                return returncode

            time.sleep(1)

        except KeyboardInterrupt:
            logger.warning('Terminating script {}'.format(script))
            process.terminate()
            process.wait()
            return process.returncode


def memoize(obj):
    """
    https://wiki.python.org/moin/PythonDecoratorLibrary#Memoize
    """
    cache = obj.cache = {}

    @functools.wraps(obj)
    def memoizer(*args, **kwargs):
        key = str(args) + str(kwargs)
        if key not in cache:
            cache[key] = obj(*args, **kwargs)
        return cache[key]

    return memoizer


def to_list(obj):
    """Convert an object to a list if it is not already one"""
    if not isinstance(obj, (list, tuple)):
        obj = [obj, ]
    return obj


def filterdict(dct, keys=None):
    """Filter dictionaries using specified keys"""
    keys = keys or []
    dct = {key: value for (key, value) in dct.iteritems() if key in keys}
    return dct


def dictproduct(dct):
    """
    Compute combinations from dictionary with list values
    by calculating the cartesian product.
    """
    # https://stackoverflow.com/questions/3873654/combinations-from-dictionary-with-list-values-using-python/41870264#41870264

    # Bring dictionary into a form suitable for computing the cartesian product
    for key, value in dct.iteritems():

        # Use empty string for empty values
        if value is None or value == [] or value == [None]:
            value = ''

        # Let's make everything a list
        value = to_list(value)

        dct[key] = value

    # Yield dictionaries making up all combinations of multiple values
    # reduced to dictionaries containing single scalar values only.
    for t in itertools.product(*dct.itervalues()):
        yield dict(zip(dct.iterkeys(), t))


class JsonObjectEncoder(JSONEncoder):
    """
    Serialize nested object compositions to JSON
    """

    def default(self, o):
        # TODO: Maybe use "o.asdict()"?
        return o.__dict__


def unique(data):
    return list(set(data))


def object_to_dictionary(obj, rules):
    """
    Transform object to dictionary according to set of rules.
    For an example, see ``p2n.formatter.tables.pivottables_data_documents``.

    TODO: Provide examples of rule format, input and output data.
    """

    data = OrderedDict()

    for rule in rules:
        key = value = None

        if isinstance(rule, types.StringType):
            key = rule
            try:
                value = operator.attrgetter(rule)(obj)
            except AttributeError:
                pass

        elif isinstance(rule, types.DictionaryType):
            key = rule['name']
            value = rule['getter'](obj)

            if 'recipe' in rule:
                recipe = rule['recipe']
                value = recipe(value)

        if key:
            data[key] = value

    return data


def exception_traceback(exc_info=None):
    """
    Return a string containing a traceback message for the given
    exc_info tuple (as returned by sys.exc_info()).

    from setuptools.tests.doctest
    """

    if not exc_info:
        exc_info = sys.exc_info()

    # Get a traceback message.
    excout = StringIO()
    exc_type, exc_val, exc_tb = exc_info
    traceback.print_exception(exc_type, exc_val, exc_tb, file=excout)
    return excout.getvalue()


@memoize
def find_convert():
    """
    Debian: aptitude install imagemagick
    /usr/bin/convert

    Mac OS X with Homebrew
    /usr/local/bin/convert

    Mac OS X with Macports
    /opt/local/bin/convert

    Self-compiled
    /opt/imagemagick/bin/convert
    /opt/imagemagick-7.0.2/bin/convert
    """

    # Some nailed location candidates
    candidates = [
        '/opt/imagemagick-7.0.2/bin/convert',
        '/opt/imagemagick/bin/convert',
        '/usr/local/bin/convert',
        '/opt/local/bin/convert',
        '/usr/bin/convert',
    ]

    # More location candidates from the system
    candidates += where.where('convert')

    # Find location of "convert" program
    convert_path = find_program_candidate(candidates)

    logger.info('Found "convert" program at {}'.format(convert_path))
    return convert_path


def find_program_candidate(candidates):
    for candidate in candidates:
        if os.path.isfile(candidate):
            return candidate


def to_png(tiff, width=None, height=None):
    """
    Convert image to PNG format with optional resizing.

    :param tiff: A stream buffer object like BytesIO
    :param width: The width of the image in pixels (optional)
    :param height: The height of the image in pixels (optional)
    :return: A BytesIO object instance containing image data
    """

    """
    The PIL module didn't properly support TIFF images with G4 compression::

        Failure: exceptions.IOError: decoder group4 not available
        Maybe patch: http://mail.python.org/pipermail/image-sig/2003-July/002354.html

    Nowadays, this should be supported by Pillow on recent platforms:
    https://pillow.readthedocs.io/en/latest/releasenotes/5.0.0.html#compressed-tiff-images
    """
#    #◘ the uglyest code in my life
#    if 'KR2019' in tiff.name or 'KR2018' in tiff.name:
#        logger.error('Image conversion using Pillow and ImageMagicks "convert" program failed: ')
#        png= BytesIO()
#        return png
#        raise AssertionError('Image conversion using Pillow and ImageMagicks')
    try:
        from PIL import Image

        # Read image
        image = Image.open(tiff)

        if width and height:

            # Convert image to grayscale
            image = image.convert('L')

            # Resize image
            image.thumbnail((width, height), Image.LANCZOS)
        png= BytesIO()
        # Save image into a stream buffer
        try:
            image.save(png, 'PNG')
        except:
            pass
        # Readers should start reading at the beginning of the stream
        png.seek(0)

        return png

    except Exception as ex:
        logger.warning('Image conversion using "Pillow" failed: {}'.format(ex))

    """
    However, if the conversion using "Pillow" fails for some reason,
    let's try to use the "convert" utility from ImageMagick.


    Instructions for installing ImageMagick on Debian::
        apt install imagemagick
    Instructions for installing ImageMagick on Windows::
        https://www.imagemagick.org/script/download.php#windows
    Instructions for building ImageMagick on Debian::
        # https://packages.debian.org/source/wheezy/imagemagick
        aptitude install build-essential checkinstall ghostscript libbz2-dev libexif-dev fftw-dev libfreetype6-dev libjasper-dev libjpeg-dev liblcms2-dev liblqr-1-0-dev libltdl-dev libpng-dev librsvg2-dev libtiff-dev libx11-dev libxext-dev libxml2-dev zlib1g-dev liblzma-dev libpango1.0-dev
        ./configure --prefix=/opt/imagemagick-7.0.2
        wget http://www.imagemagick.org/download/ImageMagick.tar.gz
        # untar and cd
        make -j6 && make install
    """

    more_args = []

    # Compute value for "resize" parameter
    size = ''
    if width or height:

        if width:
            size += str(width)

        # Use "x" for separating "width" and "height" when resizing
        size += 'x'

        if height:
            size += str(height)

        more_args += ['-resize', size]

    convert = find_convert()
    if not convert:
        message = 'Could not find ImageMagick program "convert", please install from e.g. https://imagemagick.org/'
        logger.error(message)
        raise AssertionError(message)

    command = [
        convert,
        '+set', 'date:create', '+set', 'date:modify',
        '-define', 'stream:buffer-size=0',
        '-colorspace', 'rgb', '-flatten', '-depth', '8',
        '-antialias', '-quality', '100', '-density', '300',
        # '-level', '30%,100%',

        # Debugging
        # (see "convert -list debug")
        #'-verbose',
        #'-debug', 'All',

    ] \
        + more_args + \
        [

        # Convert from specific format
        #'{0}:-'.format(format),

        # Convert from any format
        '-',

        # Convert to PNG format
        'png:-',
    ]

    command_string = ' '.join(command)
    try:
        logger.debug('Converting image using "{}"'.format(command_string))
        return run_imagemagick(command, tiff.read())

    except Exception as ex:
        logger.error('Image conversion using ImageMagicks "convert" program failed: {}'.format(ex))
        raise AssertionError(message)


def run_imagemagick(command, input=None):
    output = run_command(command, input)
    if 'ImageMagick' in output.read()[:200]:
        command_string = ' '.join(command)
        message = 'Image conversion failed, found "ImageMagick" in STDOUT. Command was "{}"'.format(
            command_string)
        logger.error(message)
        raise RuntimeError(message)
    output.seek(0)
    return output


def run_command(command, input=None):

    command_string = ' '.join(command)

    proc = subprocess.Popen(
        command,
        #shell = (os.name == 'nt'),
        #shell = True,
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )

    stdout = stderr = ''
    try:
        stdout, stderr = proc.communicate(input)
        if proc.returncode is not None and proc.returncode != 0:
            message = 'Command "{}" failed, returncode={}, stderr={}'.format(
                command_string, proc.returncode, stderr)
            logger.error(message)
            raise RuntimeError(message)

    except Exception as ex:
        if isinstance(ex, RuntimeError):
            raise
        else:
            message = 'Command "{}" failed, returncode={}, exception={}, stderr={}'.format(
                command_string, proc.returncode, ex, stderr)
            logger.error(message)
            raise RuntimeError(message)

    return BytesIO(stdout)

    """
    # Use Delegator.py for process execution

    # Currently, there seem to be problems using both binary STDIN and STDOUT:
    # https://github.com/kennethreitz/delegator.py/issues/51
    # Let's try again soon.

    #proc = delegator.run(command)
    #proc = delegator.run(command, block=True, binary=True)
    proc = delegator.run(command, block=False, binary=True)

    # https://github.com/kennethreitz/delegator.py/issues/50
    #proc.blocking = False
    proc.send(tiff.read())
    #proc.subprocess.send(tiff.read() + b"\n")
    proc.block()
    #print 'out:', proc.out
    print 'self.blocking:', proc.blocking
    print 'self._uses_subprocess:', proc._uses_subprocess
    print 'self.subprocess:', proc.subprocess

    print 'stdout-1:', proc.std_out
    stdout = proc.std_out.read()
    print 'stdout-2:', stdout
    #if 'ImageMagick' in stdout[:200]:
    #    raise ValueError('Image conversion failed, found "ImageMagick" in STDOUT')

    return BytesIO(stdout)
    """
